src/interfaces/python/swig_typemaps.i

```diff
@@ -824,71 +633,129 @@ static bool spvector_to_numpy(PyObject* &obj, SGSparseVector<type> sg_vector, in
 #endif
 
 /* One dimensional input arrays */
-%define TYPEMAP_IN_SGVECTOR(type,typecode)
-%typemap(typecheck, precedence=SWIG_TYPECHECK_POINTER) shogun::SGVector<type>
+%define TYPEMAP_SGVECTOR(type, typecode)
+
+// used for "in" typemap
+%fragment(SWIG_AsVal_frag(shogun::SGVector<type>),"header")
 {
-    $1 = is_pyvector($input, typecode);
+    int SWIG_AsVal_dec(shogun::SGVector<type>)(PyObject* input, shogun::SGVector<type>* val)
+    {
+        // if val is nullptr, it's expected to do typechecking
+        if (!val)
+            return is_pyvector(input, typecode)? SWIG_OK : SWIG_ERROR;
+        else
+            return vector_from_numpy<type>(*val, input, typecode)? SWIG_OK : SWIG_ERROR;
+    }
 }
 
-%typemap(in) shogun::SGVector<type>
+// used for "out" typemap
+%fragment(SWIG_From_frag(shogun::SGVector<type>),"header")
 {
-    if (!vector_from_numpy<type>($1, $input, typecode))
-        SWIG_fail;
+    PyObject* SWIG_From_dec(shogun::SGVector<type>)(const shogun::SGVector<type>& arg)
+    {
+        PyObject* result = nullptr;
+        vector_to_numpy(result, arg, typecode);
+        return result;
+    }
 }
-%enddef
 
-/* Define concrete examples of the TYPEMAP_IN_SGVECTOR macros */
-TYPEMAP_IN_SGVECTOR(bool,          NPY_BOOL)
-#ifdef PYTHON3 // str -> unicode for python3
-TYPEMAP_IN_SGVECTOR(char,          NPY_UNICODE)
-#else
-TYPEMAP_IN_SGVECTOR(char,          NPY_STRING)
-#endif
-TYPEMAP_IN_SGVECTOR(uint8_t,       NPY_UINT8)
-TYPEMAP_IN_SGVECTOR(int16_t,       NPY_INT16)
-TYPEMAP_IN_SGVECTOR(uint16_t,      NPY_UINT16)
-TYPEMAP_IN_SGVECTOR(int32_t,       NPY_INT32)
-TYPEMAP_IN_SGVECTOR(uint32_t,      NPY_UINT32)
-TYPEMAP_IN_SGVECTOR(int64_t,       NPY_INT64)
-TYPEMAP_IN_SGVECTOR(uint64_t,      NPY_UINT64)
-TYPEMAP_IN_SGVECTOR(float32_t,     NPY_FLOAT32)
-TYPEMAP_IN_SGVECTOR(float64_t,     NPY_FLOAT64)
-TYPEMAP_IN_SGVECTOR(floatmax_t,    NPY_LONGDOUBLE)
-TYPEMAP_IN_SGVECTOR(complex128_t,   NPY_CDOUBLE)
-TYPEMAP_IN_SGVECTOR(PyObject,      NPY_OBJECT)
-
-#undef TYPEMAP_IN_SGVECTOR
-
-/* One dimensional output arrays */
-%define TYPEMAP_OUT_SGVECTOR(type,typecode)
-%typemap(out) shogun::SGVector<type>
+// tell swig about the traits of the type to be used in stl containers
+%fragment(SWIG_Traits_frag(shogun::SGVector<type>), "header",
+          fragment="StdTraits",
+          fragment=SWIG_From_frag(shogun::SGVector<type>),
+          fragment=SWIG_AsVal_frag(shogun::SGVector<type>),
+          fragment=SWIG_From_frag(std::string),
+          fragment=SWIG_AsVal_frag(std::string))
```

Comment:
this fragment makes std::vector aware of SGVector typemaps

Reply:
pretty cool! :)